pkg/cloudprovider/azure.go

```diff
@@ -52,9 +175,202 @@ func (p *AzureProvider) Validate(secret *kapi.Secret) []string {
+	return &env, err
+}
+
+func (p *AzureProvider) newServicePrincipalToken(tenantID, clientID, clientSecret string, env *azure.Environment) (*adal.ServicePrincipalToken, error) {
+	oauthConfig, err := adal.NewOAuthConfig(env.ActiveDirectoryEndpoint, tenantID)
+	if err != nil {
+		return nil, err
+	}
+
+	return adal.NewServicePrincipalToken(
+		*oauthConfig,
+		clientID,
+		clientSecret,
+		env.ResourceManagerEndpoint,
+	)
+}
+
+func (p *AzureProvider) getStorageAccountKey(azureCreds map[string]string) (string, *azure.Environment, error) {
+	// 1. Get Azure cloud from AZURE_CLOUD_NAME, if it exists. If the env var does not
+	// exist, parseAzureEnvironment will return azure.PublicCloud.
+	env, err := p.getAzureEnvironment(azureCreds[cloudNameKey])
+	if err != nil {
+		return "", nil, err
+	}
+
+	// 2. Set Azure subscription ID
+	subscriptionID := azureCreds[subscriptionIDKey]
+
+	// 3. Get Service Principal Token (SPT)
+	spt, err := p.newServicePrincipalToken(
+		azureCreds[tenantIDKey],
+		azureCreds[clientIDKey],
+		azureCreds[clientSecretKey],
+		env,
+	)
+	if err != nil {
+		return "", env, err
+	}
+
+	// 4. Get storageAccountsClient
+	storageAccountsClient := storagemgmt.NewAccountsClientWithBaseURI(
+		env.ResourceManagerEndpoint,
+		subscriptionID,
+	)
+	storageAccountsClient.Authorizer = autorest.NewBearerAuthorizer(spt)
+
+	// 5. Get storage key
+	res, err := storageAccountsClient.ListKeys(context.TODO(), p.ResourceGroup, p.StorageAccount)
+	if err != nil {
+		return "", env, errors.WithStack(err)
+	}
+	if res.Keys == nil || len(*res.Keys) == 0 {
+		return "", env, errors.New("No storage keys found")
+	}
+
+	var storageKey string
+	for _, key := range *res.Keys {
+		// uppercase both strings for comparison because the ListKeys call returns e.g. "FULL" but
+		// the storagemgmt.Full constant in the SDK is defined as "Full".
+		if strings.ToUpper(string(key.Permissions)) == strings.ToUpper(string(storagemgmt.Full)) {
+			storageKey = *key.Value
+			break
+		}
+	}
+
+	if storageKey == "" {
+		return "", env, errors.New("No storage key with Full permissions found")
+	}
+
+	return storageKey, env, nil
+}
+
+type AzureBlobTest struct {
+	key               string
+	container         string
+	storageAccount    string
+	storageAccountKey string
+	azureEnv          *azure.Environment
+	client            *azstorage.BlobStorageClient
+}
+
+func (r *AzureBlobTest) Run() error {
+	client, err := r.getBlobClient()
+	if err != nil {
+		return err
+	}
+	r.client = client
+
+	err = r.upload()
+	if err != nil {
+		return err
+	}
+	defer r.delete()
+
+	err = r.download()
+	if err != nil {
+		return err
+	}
+
 	return nil
 }
+
+func (r *AzureBlobTest) upload() error {
+	blob, err := r.getBlob()
+	if err != nil {
+		return err
+	}
+
+	blob.CreateBlockBlobFromReader(bytes.NewReader([]byte{0}), nil)
+
+	return err
+}
+
+func (r *AzureBlobTest) download() error {
+	blob, err := r.getBlob()
+	if err != nil {
+		return err
+	}
+
+	_, err = blob.Get(nil)
```

Comment:
@dymurray `blob` in this case already holds the reference to the key from the storageContainer we want (key == filename), so we can run a `blob.Get(nil)` where nil means no Get options are specified and it will retrieve the preset key name from the preset containerName.

See `r.container`, `r.key`
```
func (r *AzureBlobTest) getBlob() (*azstorage.Blob, error) {
	container := r.client.GetContainerReference(r.container)
	if container == nil {
		return nil, errors.Errorf("unable to get container reference for bucket %v", r.container)
	}

 	blob := container.GetBlobReference(r.key)
	if blob == nil {
		return nil, errors.Errorf("unable to get blob reference for key %v", r.key)
	}

 	return blob, nil
}
```